core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -365,7 +350,7 @@ private void startNewPage()
             currentPageBuilder = currentPageBuilder.newPageBuilderLike();
```

Comment:
For fixed width type you can assume here that there will be `OUTPUT_PAGE_SIZE` positions (type check on `instaceof FixedWidthType`.
For variable length type, (`instanceof VariableWidthType`) you can assume `OUTPUT_PAGE_SIZE` positions, but can assume some skew via `io.trino.spi.block.BlockUtil#calculateBlockResetBytes`. You can then use `io.trino.spi.type.Type#createBlockBuilder(io.trino.spi.block.BlockBuilderStatus, int, int)` to construct builder of right size.

This should save memory compared to existing code. I wouldn't be surprised to see aggregation mem usage drop by `10-20%`

Reply:
Doing this properly is not that straightforward. `currentPageBuilder` is `PageBuilder` not a single `BlockBuilder` so this would have to be implemented in the `PageBuilder`.
Also, there is no direct access to the number of bytes in the `VariableWidthBlockBuilder`.We could use `getSizeInBytes` but the result contains also size of `valueIsNull` and `offsets` arrays.
IMO this looks like a separate PR. WDYT?